### DISCUSSION TITLE
Add Thomson TO16PC Emulation

### DISCUSSION THREAD

**Discussion by jvernet** [86Box/86Box#4733] [Ideas]
**Is your feature request related to a problem? Please describe.**
Add THOMSON TO16PC to 86Box
The TO16PC is a french XT compatible (8088/10Mhz, 512 to 768k of RAM, Internal graphic card: a Plantronics COLORPLUS)
The 128k extra ram can be used as LIM EMS or disk drive, altough no driver exist...
It's near the Olivetti M19 without Olivetti stuff.

**Describe the solution you'd like**
I started to add the system in my own fork, was able to boot simply replacing any XT compatible ROM with the 32K rom.

But need to add :
- SWitchs configuration
- internal graphic card (it's possible to deactivate internal card and use any other card, but the BIOS should be informed by reading SWichs.
- memory paging system.

And I failed  to add that... See my own fork of 86Box. I need help to finish. I have some informations about SWitchs configs and memory location, PPI port needed, etc.

**Additional context**
![image](https://user-images.githubusercontent.com/16312616/104358839-470f7d80-550f-11eb-846f-faa5423ca78d.png)

**Comment by ghost** [comment]
Where can I find the roms for the specific machine. I may try implementing it.

**Comment by jvernet** [comment]
Here they are. 32k. I alos some have tecnicla information about swicth, memory, memory banking.

[to16pc.zip](https://github.com/86Box/86Box/files/6054815/to16pc.zip)

**Comment by ghost** [comment]
I've found a technical manual about it. But I've found nothing related to how it treats extended memory. You can provide a manual with it freely. It'll be helpful.

**Comment by Fenix770** [comment]
Hello!

I found a manual which shows in detail the design of the internal circuitry of the machine. I has 46 pages.  
I found it in an old forum:

The here link is here on the forum:

[https://forum.system-cfg.com/viewtopic.php?t=2027&start=60](url)

And the link of the manual in PDF format: 
[http://www.lauraetjean-damien.com/system-cfg/THOMSON/COFADEL-TO16/COFADEL-TO16-MANUEL-DE-MAINTENANCE.ZIP](url)

I'm not a programmer or engineer, but the mainboard seems to have a floppy disk controller and video controller on the mainboard. So maybe the extra 128kb ram is used in some way by some video controller.

In PCEM, the BIOS of this machine can detect and differentiate between Hercules, CGA, Plantronics Color Plus and EGA video cards.
Here is the link [https://www.pcem-emulator.co.uk/phpBB3/viewtopic.php?f=5&t=2803](url)

Maybe this does not support EMS memory. Because a proper hardware controller is required. The LIM EMS 3.2 spec (which was the most common back in late '80). The LIM EMS window is 256kb. The EMS hardware does a bank switching, changing the window when required.bI read the manual of the emulated EMS board Micro Mainframe 5150T and The Laser XT 8088 which has an internal EMS hardware and according both manuals the minimum supported memory is 256kb. Additional memory should be installed in 256kb increments. (512,768,1024kb, etc).
(of course on the emulated machine any amount can be installed, but not in the real hardware).

Also, I installed the Thomson MS-DOS 3.21, on PCEM, and there is no EMM.SYS or any other EMS device driver on the Setup disk.

Here is the link in of the MS-DOS and utilities:
[https://vetusware.com/download/Thomson%20TO16%20Disquettes%20Systemes%20FR%203.21%20FR/?id=8894](url)

**Comment by ghost** [comment]
Thanks for the details but I do have all the things you gave above except the Tools diskette. According to the manual it doesn't seem to be generic EMS.

**Comment by jvernet** [comment]
It do not support EMS directly, because there is NO EMM.SYS written for this machine, nor RAMDRIVE.SYS... It has to be done !

Here is a link to some information about Switch and where they are mapped: in french
[http://www.forler.ch/misc/HWConfig_TO16.pdf](url)

More over to complete this documentation

"
Le changement du mode turbo se fait simplement en accédant, peu importe la manière, aux ports 0C0h-0C3h. Pour lire la vitesse courante, c'est les ports 0E0h-0E3h (tout est dans la doc que j'ai publiée, page 4 pour cette partie-la). P.2 de la même doc, la batterie SW2 est lisible à travers le PPI, ce qui est un mécanisme standard sur PC-AT : les bits 0-3 du port 62h forment 2 nibbles que l'on sélectionne à travers le bit 3 du port 61h. Au passage, le bit 4 de ce dernier active la parité RAM.

Concernant le mécanisme de commutation des 128k, tout est dans ma doc, sauf ceci : la commutation de page se fait au travers du port 208h. Si le bit7 est fixé à 1, la banque est visible. Le no de page, de 0 à 7 est codé sur les bits 0-2."

**Comment by ghost** [comment]
Alright I see it has 7 pages which are controller by Port 208h. Also I see it uses AT Keyboard mechanism. I'll read the hardware datasheet and see more.

Edit: Seems like jumpers for like the Internal video won't be needed. However I see jumpers for the Serial ports.

**Comment by ghost** [comment]
Is there any other way i can obtain the diskettes for testing?

**Comment by Fenix770** [comment]
Here is the TO16 MS-DOS 3.21. It seems to be a rather standard DOS.
[Thomson TO16 Systems Disks FR.zip](https://github.com/86Box/86Box/files/6069802/Thomson.TO16.Systems.Disks.FR.zip)

And a MFM 20mb Hard Disk: With TO16 MS-DOS installed: (C/H/S 615/4/17).
[Thomson TO16.zip](https://github.com/86Box/86Box/files/6069817/Thomson.TO16.zip)

It also includes a mouse driver MOUSE.COM and MOUSE.SYS and SLOW.COM and FAST.COM utilities, I think it is to change clock speed. (Altough In PCem does nothing)

**Comment by ghost** [comment]
Thank you. I checked PCems but PCem does not emulate the TO16 properly (lacks everything pretty much)

**Comment by ghost** [comment]
I may have to set that machine on hold. Talked to OBat and it surely needs an EMS API in order for us to continue.

**Comment by jvernet** [comment]
There is no known EMM.SYS driver for this machine. Need to be writen, may be starting with the Lotech Driver Sources ?

At least, implement the configuration switch setting and reading, so that the BIOS who heavily read these will at least start with no error (beep),set the internal graphic card to Color Plus by default.

**Comment by OBattler** [comment]
By EMS API, he means an EMS API on the emulator's side to make implementing chipset EMS support easier. I have that in my plans.

**Comment by jvernet** [comment]
A little update to this request: the TO16 use the same chipset ( (MT86G011), and logic glue (MT86G16A /MT86M001A) as the ACER 710 .
If this can help.

**Comment by OBattler** [comment]
It does help - the chipset is basically all the XT stuff on one chip + EMS.

**Comment by OBattler** [comment]
OK, so the upper memory is intended for UMB's, not EMS. So, if the third bank is enabled, you get 768k RAM, of which, the top most 128k are used for the banked memory (C0000-EC000).

**Reply by bozoscum** [reply]
According to the documents mentioned in this discussion thread and the BIOS code , the 'Extension RAM' is actually been emulated as EMS

**Comment by Fenix770** [comment]
I would be interesting to see if the RAMDRIVE.SYS included on the Thomson DOS 3.21 can use that extra memory.

**Comment by OBattler** [comment]
Also, what RTC does this use?

**Comment by flama12333** [comment]
MM58167AN https://forum.system-cfg.com/viewtopic.php?p=178866#p178866 
XP Model only

**Comment by bozoscum** [comment]
<img width="722" height="497" alt="image" src="https://github.com/user-attachments/assets/acfb595a-86e0-401b-9b93-aa1cf78ed313" />

[BIOS Extension RAM check routine](https://github.com/user-attachments/files/23766612/TO16_DMP.TXT)

from the BIOS routine dumped above, the RAM above 640k can be access from 16KB memory window whose address is specified by the lower 4bit of SW3 with maximum 24 pages switch via port 208h, and maximum of 128 pages switched via port 218h
my wild guess is the the port 218h is for extension memory card up to 2MB size

as for the RTC, I have implemented and it works like a charm. The BIOS will take the Date/Time from RTC.

**Reply by OBattler** [reply]
I wonder how difficult would it be to implement those EMS pages. 24 pages would be 24 * 16 = 384 kB, so I guess that handles remapping the on-board upper 384 kB, and yes, 128 pages on port 218h would be 2 MB. I wonder how the entire interface looks like - do the two ports use separate EMS windows or are they switched? And if yes, how?

**Reply by bozoscum** [reply]
I can see from the BIOS routine is that the two ports use the same EMS window.

<img width="516" height="760" alt="image" src="https://github.com/user-attachments/assets/779e1cfd-ab13-4149-b5f5-c84f4eb475ab" />

the subroutine DE39 get the EMS window base address from DIP SW3 (actually, read from port 0E0h, which is machine status port of ALi M1101 chipset) and is called only once. Page number is assigned in the lower 7 bits of port 208h/218h, and the highest bit seems to be activating flag just as the document HWConfig_TO16.pdf said.

I don't think it is very difficult to implement the pages allocated between 640KB to 1MB but first I need to figure out the current memory paging mechanism implemented in 86Box

**Reply by OBattler** [reply]
You need to disable ram_mid_mapping, and add your own mapping for that page, with configurable address, then the port 208h handler specifies what physical address it maps to - see eg. the Laser XT code, how EMS is implemented there.

**Reply by bozoscum** [reply]
thanks for the hint, it saves lot of my time
